views/reset.go

```diff
@@ -0,0 +1,137 @@
+		}
+		p := c.Request().Form.Get("password")
+		if p != c.Request().Form.Get("confirmpassword") || p == "" {
+			return v.template.RenderTemplate(c.Response(), nil, templates.ResetTemplate, templates.NoNavType)
+		}
+		user1.Password = null.StringFrom(password)
+
+		user2, err := v.user.UpdateUserPassword(c.Request().Context(), user1)
+		if err != nil {
+			log.Printf("failed to reset user: %+v", err)
+		}
+		v.cache.Delete(url)
+		log.Printf("updated user: %s", user2.Username)
+		return c.Redirect(http.StatusFound, "/")
+	}
+	return nil
+}
+
+func (v *Views) ResetUserPasswordFunc(c echo.Context) error {
+	session, _ := v.cookie.Get(c.Request(), v.conf.SessionCookieName)
+
+	c1 := v.getData(session)
+
+	userID, err := strconv.Atoi(c.Param("userid"))
+	if err != nil {
+		return echo.NewHTTPError(http.StatusInternalServerError, fmt.Errorf("failed to parse userid for reset: %w", err))
+	}
+
+	user1, err := v.user.GetUser(c.Request().Context(), user.User{UserID: userID})
+	if err != nil {
+		return echo.NewHTTPError(http.StatusInternalServerError, fmt.Errorf("failed to get user for reset: %w", err))
+	}
+
+	user1.ResetPw = true
+
+	_, err = v.user.UpdateUser(c.Request().Context(), user1, c1.User.UserID)
+	if err != nil {
+		return echo.NewHTTPError(http.StatusInternalServerError, fmt.Errorf("failed to update user for reset: %w", err))
+	}
+
+	url := uuid.NewString()
+	v.cache.Set(url, user1.UserID, cache.DefaultExpiration)
+
+	var status int
+
+	var message struct {
+		Message string `json:"message"`
+		Error   error  `json:"error"`
+	}
+
+	// Valid request, send email with reset code
+	if v.mailer != nil {
+		v.mailer = mail.NewMailer(mail.Config{
+			Host:       v.conf.Mail.Host,
+			Port:       v.conf.Mail.Port,
+			Username:   v.conf.Mail.Username,
+			Password:   v.conf.Mail.Password,
+			DomainName: v.conf.DomainName,
+		})
+
+		file := mail.Mail{
+			Subject: "YSTV Security - Reset Password",
+			Tpl:     v.template.RenderEmail(templates.ResetEmailTemplate),
+			To:      user1.Email,
+			From:    "YSTV Security <no-reply@ystv.co.uk>",
+			TplData: struct {
+				Email string
+				URL   string
+			}{
+				Email: user1.Email,
+				URL:   fmt.Sprintf("https://%s/reset/%s", v.conf.DomainName, url),
+			},
+		}
+
+		err = v.mailer.SendMail(file)
+		if err != nil {
+			return echo.NewHTTPError(http.StatusInternalServerError, fmt.Errorf("failed to send email for login: %w", err))
+		}
+
+		log.Printf("request for password reset email: \"%s\"", user1.Email)
+		message.Message = fmt.Sprintf("Reset email sent to: \"%s\"", user1.Email)
+	} else {
+		message.Message = fmt.Sprintf("No mailer present\nPlease forward the link to this email: %s, reset link: https://%s/reset/%s", user1.Email, v.conf.DomainName, url)
+		message.Error = fmt.Errorf("no mailer present")
+		log.Printf("no Mailer present")
+		log.Printf("reset email: %s, url: %s, reset link: https://%s/reset/%s", user1.Email, url, v.conf.DomainName, url)
+	}
```

Comment:
NITPICK: This is fairly similar to  `forgot.go`, would there be value in refactoring this out into a common function (name suggestion: `sendEmailIfConfigured` but feel free to change)

Reply:
I will do that in a bit